x-pack/plugins/lens/public/indexpattern_datasource/operations/definitions/filters/filters.tsx

```diff
@@ -0,0 +1,282 @@
+  EuiButtonIcon,
+  EuiLink,
+  EuiText,
+  EuiButtonEmpty,
+  EuiForm,
+  EuiFormRow,
+} from '@elastic/eui';
+import { updateColumnParam } from '../../../state_helpers';
+import { OperationDefinition } from '../index';
+import { FieldBasedIndexPatternColumn } from '../column_types';
+import { FilterPopover } from './filter_popover';
+import { DataType } from '../../../../types';
+
+export enum SEARCH_QUERY_LANGUAGE {
+  KUERY = 'kuery',
+  LUCENE = 'lucene',
+}
+
+export interface Filter {
+  input: Query;
+  label?: string;
+}
+
+export const emptyFilter: Filter = {
+  input: {
+    query: '',
+    language: SEARCH_QUERY_LANGUAGE.KUERY,
+  },
+  label: '',
+};
+
+const searchQueryLabel = i18n.translate('xpack.lens.indexPattern.searchQuery', {
+  defaultMessage: 'Search query',
+});
+
+const defaultSearchQuery: Filter = {
+  input: {
+    query: '*',
+    language: 'kuery',
+  },
+  label: i18n.translate('xpack.lens.indexPattern.searchQueryDefault', {
+    defaultMessage: 'All records',
+  }),
+};
+
+const countDuplicates = (filterArr: Filter[], filter: Filter) =>
+  filterArr.filter(
+    (f) =>
+      JSON.stringify(f.input.query.trim()) === JSON.stringify(filter.input.query.trim()) &&
+      f.label === filter.label
+  ).length;
+
+const makeUniqueLabel = (filterArr: Filter[], filter: Filter) => {
+  let label = '';
+  let count = 0;
+  do {
+    count++;
+    label = `${filter.input.query} [${count}]`;
+  } while (filterArr.find((f) => f.label === label));
+  return label;
+};
+
+interface DraggableLocation {
+  droppableId: string;
+  index: number;
+}
+
+export interface FiltersIndexPatternColumn extends FieldBasedIndexPatternColumn {
+  operationType: 'filters';
+  params: {
+    filters: Filter[];
+  };
+}
+
+export const filtersOperation: OperationDefinition<FiltersIndexPatternColumn> = {
+  type: 'filters',
+  displayName: searchQueryLabel,
+  priority: 3, // Higher than any metric
+  getPossibleOperationForField: ({ type }) => {
+    if (type === 'document') {
+      return {
+        dataType: 'number',
+        isBucketed: true,
+        scale: 'ordinal',
+      };
+    }
+  },
+  isTransferable: () => false,
+
+  buildColumn({ suggestedPriority, field }) {
+    return {
+      label: searchQueryLabel,
+      dataType: field.type as DataType,
+      operationType: 'filters',
+      scale: 'ordinal',
+      suggestedPriority,
+      isBucketed: true,
+      sourceField: field.name,
+      params: {
+        filters: [],
+      },
+    };
+  },
+
+  toEsAggsConfig: (column, columnId) => ({
+    id: columnId,
+    enabled: true,
+    type: 'filters',
+    schema: 'segment',
+    params: {
+      filters: column.params.filters.length > 0 ? column.params.filters : [defaultSearchQuery],
+    },
+  }),
+  onFieldChange: (oldColumn, indexPattern, field) => {
+    return {
+      ...oldColumn,
+      label: field.displayName,
+      sourceField: field.name,
+    };
+  },
+  paramEditor: ({ state, setState, currentColumn, layerId }) => {
```

Comment:
I recommend extracting this into a function component so it can be tested separately, and also so that we can `useState`